torch/_subclasses/fake_tensor.py

```diff
@@ -2522,21 +2523,61 @@ def wrap(e: T) -> Union[T, FakeTensor]:
 
         return tree_map(wrap, r)
 
-    def create_symbolic_nested_int(
-        self, *, nt_tensor_id: Optional[int] = None
+    def get_nested_int(
+        self,
+        *,
+        cache: CachedTensor,
+        coeff: Union[int, torch.SymInt] = 1,
     ) -> torch.SymInt:
+        # Note [ Best effort SymInt association ]
+        #
+        # The question this note tries to answer is why is it okay to be
+        # "best effort/unreliable" when it comes to associating raggedness with
+        # SymInts. What we mean by "unreliable" here is that in an ideal world
+        # symbolic shapes guards should serve as the perfect complement to dynamo
+        # guards, e.g., two SymInts share the same symbol if dynamo can already guard
+        # based on tensor deduping guards that they are the same. Instead of doing that,
+        # we instead associate SymInt on CachedTensor, an object that is somewhat
+        # ephemeral/disposable, e.g. the same CachedTensor gets preserved when
+        # we do something like pointwise, but calling from_jagged(values, offsets)
+        # actually always creates a new CachedTensor, and we would have a fresh
+        # SymInt in that case.
+        #
+        # The answer is that:
+        # - We don't /need/ symbolic shapes guards because we rely on Dynamo guards.
+        #   Hence, we use ephmeral sources when convenient. (There are some known
+        #   downsides: (1) actually trying to produce guards with those SymInts would
+        #   error out. See test_in_graph_construction_mixed_4
+        #   (2) we sometimes secretly update the tensor registry so that
+        #   two j0 compare equal despite not having caches that share the same tensor
+        #   instance. This is not sound, but should be rare - only happens in the .to()
+        #   case).
+        # - We don't /need/ a cache because SymInts in this model are fungible.
+        #   (All have ephemeral sources.)
+        # - All we care about is that the common case (pointwise) doesn't
+        #   create a new symbol/SymInt everytime.
+        # - It's okay to be best effort/not be reliable because the caching is just
+        #   a optimization that doesn't matter for correctness.
```

Comment:
and on top of that - is it fair to say that in any cases where we are sub-optimal because we failed to realize two nested_ints are the same, this is always fixable with some user code changes?

Reply:
> we failed to realize two nested_ints are the same, this is always fixable with some user code changes?

Yeah:
- users aren't always able to make their offsets the same instance (solvable via code changes)
- we aren't always able to make the fact that offsets are the same instance statically known to the compiler (via making symbols the same) thus incurring unnecessary guards.